### ISSUE TITLE
AutoGluon for very small datasets?

### ISSUE THREAD

**Issue by sergeyf** [awslabs/autogluon#833]
Hello,

Thanks for the hard work on this package! 

I was wondering, do the authors/maintainers of this package have experience with applying AutoGluon to very small datasets? For example, N = 100, or N = 500. Especially in settings where the dimensionality D > N.

For such small datasets, my experience is that Ridge Regression or Elastic Net are hard to beat because it's too easy to overfit to a tiny validation set while trying a lot of models or a lot of hyperparameters. That is, it's easy to find a model that does well on the validation set by accident and this model won't generalize to the test set.

Does your experience suggest that autogluon's particular approach is able to avoid such pitfalls? If so, can you point me to case-studies?

Thank you.

**Comment by Innixma** [comment]
Thanks for the interesting question!

Which of the tasks in AutoGluon are you referring to?

If you are talking primarily about Tabular, then I believe AutoGluon will do well in small dataset scenarios when repeated bagging is enabled. We don't actually do hyperparameter tuning in Tabular by default, which helps immensely to mitigate overfitting. Furthermore, repeated bagging will help stabilize the result and improve consistency and overall accuracy.

The primary area where AutoGluon might struggle is overfitting via early stopping. In v0.1 release planned in early February, we will have options for the user to specify how early stopping is performed (or if it is used at all) which may help in this regard.

I'd recommend giving AutoGluon a try, and then if you find other techniques work better consistently in small dataset scenarios, it would be helpful to respond back here to see if we can integrate the technique into AutoGluon.

**Comment by sergeyf** [comment]
Thanks for your answer @Innixma. Yes, I was referring to tabular regression/classification.

I have a half-formed plan to apply AutoGluon to a bundle of small UCI datasets and see how it fares against (a) my favorite linear models and (b) just LightGBM which I have had success in tuning properly. I'll certainly report back.

**Comment by sergeyf** [comment]
Hello again. I got 142 UCI datasets that someone else packaged up and ran some baselines on them. You get a plot like this (each dot is a dataset):

![image](https://user-images.githubusercontent.com/1874668/103058574-5a09e000-4557-11eb-9f71-8c63524fc158.png)

My goal is to add an AutoGluon line. What's a reasonable time budget for each dataset? The baseline models run in ~30s to 10m.

**Comment by jwmueller** [comment]
Note that you don't need to specify a time budget to use AutoGluon. However, if you find it is taking too long, then you could set a time budget based on how long you're willing to wait (or perhaps 1-10x as long as your most intensive model since AutoGluon performs better when it is able to train many models). 

If you care most about obtaining the best possible AUROC, make sure to run autogluon's  `fit()` with arguments: 
`presets='best_quality'`, `eval_metric='roc_auc'`
as detailed here: 
https://auto.gluon.ai/stable/tutorials/tabular_prediction/tabular-quickstart.html#maximizing-predictive-performance

**Comment by sergeyf** [comment]
Thanks!  I dd want to use a time budget because I was doing 4-fold nested cross-validation on 142 datasets, so a budget of 60s was equivalent to 9.333 hours. I also ran a 120s budget. Here are the results.

Number of datasets each algorithm does best on:
Counter({'AutoGluon (sec=120)': 84, 'AutoGluon (sec=60)': 74, 'LightGBM (n_iter=25)': 74, 'LightGBM (n_iter=10)': 68, 'Logistic Regression': 64, 'Random Forest': 64, 'SVC': 35}) 
```python
Average AUROC performance for each algorithm:
AutoGluon (sec=120)     0.887491
AutoGluon (sec=60)      0.886326
LightGBM (n_iter=10)    0.886359
LightGBM (n_iter=25)    0.888885
Logistic Regression     0.857380
Random Forest           0.886253
SVC                     0.852368

Median AUROC performance for each algorithm:
AutoGluon (sec=120)     0.924359
AutoGluon (sec=60)      0.925754
LightGBM (n_iter=10)    0.924920
LightGBM (n_iter=25)    0.927121
Logistic Regression     0.891617
Random Forest           0.922966
SVC                     0.880665
```

And a plot:
![image](https://user-images.githubusercontent.com/1874668/103418635-62a87a80-4b44-11eb-8d8e-0fedc46575aa.png)

Also a plot for performance vs size of dataset:
![image](https://user-images.githubusercontent.com/1874668/103418659-81a70c80-4b44-11eb-8e0c-5088334703e5.png)

My conclusion is that maybe the time budget is too small for the datasets closer to 10k. But, broadly, I'm not seeing wins over just using LightGBM with 25 random hyperparameter evaluations.

It's all here if you're interested: https://github.com/sergeyf/SmallDataBenchmarks/ 

Comments/criticisms are most welcome

**Comment by Innixma** [comment]
120s is definitely on the very low end of time limits and isn't enough to realistically estimate how powerful AutoGluon (or any AutoML system) is. For our benchmarking we usually run for 1 hour or 4 hours on `presets='best_quality'`. I notice in your code you have not enabled any of the presets, meaning AutoGluon is on `'medium_quality'`. This will result in worse scores (although 120s is likely not enough time for `'best_quality'`).

For additional context, refer to our [paper](https://arxiv.org/pdf/2003.06505.pdf).

In Table S5, you will see a comparison of different settings in AutoGluon. The one you used (medium_quality) is equal to the `NoBag` framework in the table. `'best_quality'` is named `AutoGluon`. This also shows that time limits beyond 1 hour continue to show significant improvement.

**Comment by sergeyf** [comment]
Thanks @Innixma, I appreciate your pointers.

I will change the preset to `best_quality` and rerun with a 5m budget. I understand that may not be enough to achieve "best" performance, but this comparison is designed to compare "roughly equivalent" algorithms on tiny datasets, and allocating 1 hour feels like an unfair advantage when LightGBM hyperopt search runs in less than 1m. It would, however, be nice to see an upper bound to performance and whether AutoGluon would actually be best in every case or not.

However, I don't have the CPU cycles available to run it for 568 hours. :)  Let me know if you are or anyone else is interested in contributing the benchmark on some Amazon hardware! The code and dataset are publically available, so it should be very easy to do if you have access to compute.

**Comment by sergeyf** [comment]
Hi again. I reran everything with 5m, and it seemed this was enough for the smaller datasets. AutoGluon is now the best (almost) across the board, except at the very end where it probably needs more time:

![image](https://user-images.githubusercontent.com/1874668/103469230-f16cf100-4d16-11eb-8752-a0025525d728.png)

And some numbers: 

```
Number of datasets each algorithm does best on (out of 108):
Counter({'AutoGluon (sec=300)': 74,
         'LightGBM (n_hyperparams=25)': 43,
         'Random Forest': 36,
         'Logistic Regression': 27,
         'SVC': 22})

Number of datasets each algorithm does worst on (out of 108):
Counter({'Logistic Regression': 56,
         'SVC': 47,
         'Random Forest': 26,
         'LightGBM (n_hyperparams=25)': 23,
         'AutoGluon (sec=300)': 13})

Average AUROC for each model
AutoGluon (sec=300)            0.885323
LightGBM (n_hyperparams=25)    0.872977
Random Forest                  0.869753
SVC                            0.840538
Logistic Regression            0.831857

Median AUROC for each model
AutoGluon (sec=300)            0.925830
LightGBM (n_hyperparams=25)    0.915428
Random Forest                  0.908864
SVC                            0.884252
Logistic Regression            0.882680
```

There are still some datasets where AutoGluon, Logistic Regression and Random Forest fail catastrophically, but SVC or LightGBM do not. 

Thanks for engaging with me and feel free to close this issue if you'd like.

**Comment by Innixma** [comment]
Hi @sergeyf, sorry for the late response (was on vacation). Great to hear that you were able to get good results with AutoGluon! I also saw your reddit post, which was very nicely written!

Like you had mentioned, I'd be very interested in a run with more time allocated to AutoGluon, so I might come back to this dataset suite in future, possibly by integrating it with automlbenchmark which would allow for distributed runs to minimize the total runtime. Additionally, the upcoming v0.1 release of AutoGluon is much stronger than our previous release, with a 68% winrate vs v0.0.15 on automlbenchmark's 104 datasets. It would be interesting to see if these improvements also carry over to the small datasets.

**Comment by sergeyf** [comment]
Ah no worries - vacation are important :)

Nice to hear about the improvements. What do you attribute the better performance to?

**Comment by Innixma** [comment]
Improved performance comes primarily from:

1. New Model: XGBoost
2. New Model: FastAI Tabular Neural Network
3. Improved Hyperparameters: LightGBM (halved learning rate)
4. Improved Hyperparameters: CatBoost (halved learning rate)
5. Improved Neural Network handling of Categorical features
6. Improved training priority of models to incentivize most impactful models to train first, especially important with low time
7. Overall stability improvements to ensure fewer model failures internally (Specifically focusing on rare categoricals / missing values).

**Comment by sergeyf** [comment]
Sounds great, thanks for the thorough answer!

**Comment by Innixma** [comment]
@santarabantoosoo In this case you would predict with all 10 folds and average the result for the final prediction.

You would also be well served to do some basic calibration on the validation prediction probabilities if your metric is F1: AutoGluon doesn't automatically calibrate predictions to maximize F1, but you can do so by adjusting the decision boundary (by default it is 0.5, which isn't optimal for F1).

**Comment by santarabantoosoo** [comment]
Thanks a lot. 

Will do that

**Comment by santarabantoosoo** [comment]
and in my case, do I need to tinker with `num_bag_folds`, `num_bag_sets` and `num_stack_levels`, or `best_	quality` would be enough?
